src/Stratis.Features.FederatedPeg/Wallet/MultiSigTransactions.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+
+namespace Stratis.Features.FederatedPeg.Wallet
+{
+    public class MultiSigTransactions : ICollection<TransactionData>, ITransactionDataObserver
+    {
+        private readonly object lockObject;
+        private Dictionary<OutPoint, TransactionData> transactionDict;
+        private Dictionary<OutPoint, TransactionData> spendableTransactionDict;
+        private Dictionary<uint256, List<TransactionData>> withdrawalsByDepositDict;
+        private SortedDictionary<int, List<TransactionData>> spentTransactionsByHeightDict;
+
+        public int Count => this.transactionDict.Count;
+
+        public bool IsReadOnly => false;
+
+        public MultiSigTransactions()
+        {
+            this.lockObject = new object();
+            this.transactionDict = new Dictionary<OutPoint, TransactionData>();
+            this.spendableTransactionDict = new Dictionary<OutPoint, TransactionData>();
+            this.withdrawalsByDepositDict = new Dictionary<uint256, List<TransactionData>>();
+            this.spentTransactionsByHeightDict = new SortedDictionary<int, List<TransactionData>>();
+        }
+
+        private bool TryAddWithdrawal(TransactionData transactionData)
+        {
+            uint256 matchingDepositId = transactionData.SpendingDetails?.WithdrawalDetails?.MatchingDepositId;
+
+            if (matchingDepositId == null)
+                return false;
```

Comment:
exit log missing

Reply:
This class is too low-level (and the method is private), for logging. It will just slow things down.

Reply:
Maybe NoTrace on it then?

Reply:
Done.